src/mongocxx/test/collection.cpp

```diff
@@ -2808,4 +2812,141 @@ TEST_CASE("Ensure that the WriteConcernError 'errInfo' object is propagated", "[
     REQUIRE(contains_err_info);
 }
 
+TEST_CASE("expose writeErrors[].errInfo", "[collection]") {
+
+    // A helper for checking that an error document is well-formed according to our requirements:
+    auto writeErrors_well_formed = [](const bsoncxx::document::view& reply_view) -> bool {
+        if (!reply_view["writeErrors"]) {
+            return false;
+        }
+
+        const auto& errdoc = reply_view["writeErrors"][0];
+
+        auto error_code = errdoc["code"].get_int32();
+
+        // The code should always be 121 (DocumentValidationFailure):
+        if (121 != error_code) {
+            std::ostringstream os;
+            os << "writeErrors expected to have code 121, but had " << error_code << " instead";
+            throw std::runtime_error(os.str());
+        }
+
+        // We require the "details" field be present:
+        if (!errdoc["errInfo"]["details"]) {
+            throw std::runtime_error("no \"details\" field in \"writeErrors\"");
+        }
+
+        return true;
+    };
+
+    // Set up our test environment:
+    instance::current();
+
+    auto client_opts = mongocxx::options::client();
+
+    // Listen to the insertion-failed event: we want to get a copy of the server's
+    // response so that we can compare it to the thrown exception later:
+    apm_opts.on_command_succeeded(
+        [&writeErrors_well_formed](const mongocxx::events::command_succeeded_event& ev) {
+            if(!writeErrors_well_formed(ev.reply()))
+        });
+
+    client_opts.apm_opts(apm_opts);
+
+    auto mongodb_client = mongocxx::client(uri{}, client_opts);
+
+    database db = mongodb_client["prose_test_expose_details"];
+
+    const std::string collname{"mongo_cxx_driver-expose_details"};
+
+    // Drop the existing collection, if any:
+    db[collname].drop();
+
+    // Make a new collection with validation checking:
+    collection coll = db.create_collection(
+        collname,
+        make_document(kvp("validator",
+                          make_document(kvp("field_x", make_document(kvp("$type", "string")))))));
+
+    SECTION("insert a valid document") {
+        bsoncxx::builder::basic::document entry;
+
+        entry.append(kvp("_id", bsoncxx::oid()), kvp("field_x", "value0"));
+
+        REQUIRE_NOTHROW(coll.insert_one(entry.view()));
+    }
+
+    SECTION("cause a type violation on insert") {
+        bsoncxx::builder::basic::document entry;
+
+        entry.append(kvp("_id", bsoncxx::oid()), kvp("field_x", 42));
+
+        try {
+            coll.insert_one(entry.view());
+
+            // We should not make it here:
+            CHECK(false);
+        } catch (const operation_exception& e) {
+            auto rse = e.raw_server_error();
+
+            CHECK(rse.has_value());
```

Comment:
```suggestion
            CHECK(rse);
```

Some Evergreen tasks ([example](https://spruce.mongodb.com/task/cxx_driver_arm_ubuntu1804_50_compile_and_test_with_shared_libs_patch_a7be2f22f2cd4c8664e48abae4f78b6f1c9d6e99_621816e77742ae51fc2669a0_22_02_24_23_38_16/logs?execution=0)) do not recognize has_value:

```
[2022/02/24 23:51:53.108] /data/mci/a42c20c4517b667a13c1bc365b509680/mongo-cxx-driver/src/mongocxx/test/collection.cpp:2896:23: error: 'struct core::v1::optional<bsoncxx::v_noabi::document::value>' has no member named 'has_value'; did you mean 'value'?
[2022/02/24 23:51:53.108]              CHECK(rse.has_value());
[2022/02/24 23:51:53.108]                        ^
[2022/02/24 23:51:53.108] In file included from /data/mci/a42c20c4517b667a13c1bc365b509680/mongo-cxx-driver/src/mongocxx/../bsoncxx/test_util/catch.hh:17:0,
[2022/02/24 23:51:53.108]                  from /data/mci/a42c20c4517b667a13c1bc365b509680/mongo-cxx-driver/src/mongocxx/test/collection.cpp:27:
[2022/02/24 23:51:53.108] /data/mci/a42c20c4517b667a13c1bc365b509680/mongo-cxx-driver/src/mongocxx/test/collection.cpp:2896:23: error: 'struct core::v1::optional<bsoncxx::v_noabi::document::value>' has no member named 'has_value'; did you mean 'value'?
[2022/02/24 23:51:53.108]              CHECK(rse.has_value());
[2022/02/24 23:51:53.108]                        ^
[2022/02/24 23:51:53.108] /data/mci/a42c20c4517b667a13c1bc365b509680/mongo-cxx-driver/src/mongocxx/test/collection.cpp:2896:23: error: 'struct core::v1::optional<bsoncxx::v_noabi::document::value>' has no member named 'has_value'; did you mean 'value'?
[2022/02/24 23:51:53.108]              CHECK(rse.has_value());
[2022/02/24 23:51:53.108]     
```

Reply:
Ah, shoot, that's because we're using C++11! My local build turns on C++17 (it looks like the reason is that the cxx::* support for whatever reason doesn't compile on my system with:

cmake -DCMAKE_BUILD_TYPE=Debug -DLIBBSON-1.0_DIR=~/work/c-driver-install/lib/cmake/libbson-1.0/ -DCMAKE_INSTALL_PREFIX=~/work/cxx-driver-install -DCMAKE_PREFIX_PATH="~/work/c-driver-install;~/work/cxx-driver-install" -DCMAKE_CXX_STANDARD=11 -DCMAKE_C_COMPILER="/usr/bin/gcc-11" -DCMAKE_CXX_COMPILER="/usr/bin/g++-11" ../